pkg/cri/resource-manager/policy/builtin/podpools/podpools-policy.go

```diff
@@ -0,0 +1,694 @@
+	if len(pools) == 0 {
+		return nil
+	}
+	return pools[0]
+}
+
+// allocatePool returns a pool allocated for a pod.
+func (p *podpools) allocatePool(pod cache.Pod) *Pool {
+	if pool := p.allocatedPool(pod); pool != nil {
+		return pool
+	}
+	poolDef := p.getPoolDef(pod)
+	if poolDef == nil {
+		return nil
+	}
+	// Try to find a suitable pool and allocate it for the pod.
+	pools := filterPools(p.pools,
+		func(pl *Pool) bool {
+			return poolDef.Name == pl.Def.Name && (pl.Def.MaxPods > len(pl.PodIDs) || pl.Def.MaxPods == 0)
+		})
+	// Sort pools according to pool type fill order so that the
+	// first pool in the list is the preferred one.
+	switch poolDef.FillOrder {
+	case FillBalanced:
+		sort.Slice(pools, func(i, j int) bool {
+			return len(pools[i].PodIDs) < len(pools[j].PodIDs)
+		})
+	case FillPacked:
+		sort.Slice(pools, func(i, j int) bool {
+			return len(pools[i].PodIDs) > len(pools[j].PodIDs)
+		})
+	case FillFirstFree:
+		// FirstFree is already the first of the pools list.
+	}
+	if len(pools) == 0 {
+		log.Warn("cannot find free %q pool for pod %q", poolDef.Name, pod.GetName())
+		return nil
+	}
+	// Found a suitable pool. Allocate it for the pod.
+	podID := pod.GetID()
+	pool := pools[0]
+	pool.PodIDs[podID] = []string{}
+	log.Debug("allocated pool %s[%d] for pod %q", pool.Def.Name, pool.Instance, pod.GetName())
+	return pool
+}
+
+// containersInPool returns the number of containers of a pod in a pool.
+func (p *podpools) containersInPool(pod cache.Pod, pool *Pool) int {
+	if cnts, ok := pool.PodIDs[pod.GetID()]; ok {
+		return len(cnts)
+	}
+	return 0
+}
+
+// freePool removes an empty pod from a pool
+func (p *podpools) freePool(pod cache.Pod, pool *Pool) {
+	podID := pod.GetID()
+	delete(pool.PodIDs, podID)
+	delete(p.podMaxMilliCPU, podID)
+}
+
+// trackPodCPU keeps track on pod's CPU requests.
+func (p *podpools) trackPodCPU(pod cache.Pod, pool *Pool) {
+	podID := pod.GetID()
+	current := p.getPodMilliCPU(podID)
+	if max, ok := p.podMaxMilliCPU[podID]; ok {
+		if max < current {
+			p.podMaxMilliCPU[podID] = current
+		}
+	} else {
+		p.podMaxMilliCPU[podID] = current
+	}
+	// Check overbooking
+	if cpuAvail := p.availableMilliCPUs(pool); cpuAvail < 0 {
+		log.Error("overbooked pool %q, cpuset:%s: %dm / %dm CPUs used, %d mCPU available", pool.PrettyName(), pool.CPUs, pool.CPUs.Size()*1000-int(cpuAvail), pool.CPUs.Size()*1000, cpuAvail)
+	}
+}
+
+// validatePodCPU compares max CPU requests against pool CPU capacity per pod.
+func (p *podpools) validatePodCPU(pod cache.Pod, pool *Pool) {
+	// Log pod configuration error if a pool had fixed amount of
+	// CPUs per, pod but the pod failed to request the correct
+	// amount.
+	podID := pod.GetID()
+	if podmCPU, ok := p.podMaxMilliCPU[podID]; ok {
+		if pool.Def.MaxPods > 0 {
+			poolmCPUperPod := int64(pool.CPUs.Size() * 1000 / pool.Def.MaxPods)
+			mCPUerr := podmCPU - poolmCPUperPod
+			if mCPUerr < -podMilliCPUErrorMargin || mCPUerr > podMilliCPUErrorMargin {
```

Comment:
Is the idea here that somehow slight over- and undersubscriptions will even/cancel each other out, or why do we allow both ?

Reply:
The purpose of this margin is to allow rounding errors in the magnitude of 0.5 mCPU/container up to 20 containers/pod - and allow the error in rounding up and down. (Therefore both over and undersubscription.)

Reply:
Could we then spell that out explicitly with a comment ?

Reply:
Clarifying comments added here and next to declaring the constant `podMilliCPUErrorMargin`.